EOS-24057: Creation of motr and s3server environment files in miniprovisioner

Shreya-18 opened the pull request:
### Description:
At present Motr and S3server environment files are created in hctl bootstrap. In the container environment, hctl bootstrap won't be running, so need to create the environment files in hare mini provisioner- config stage.

### Solution:
Generating the sysconfig files for m0d and s3server.

supriyachavan4398 commented:
After running `make` and `make devinstall` :
```
[root@ssc-vm-g2-rhev4-0399 ~]# cd /opt/seagate/cortx/hare/bin/

[root@ssc-vm-g2-rhev4-0399 bin]# ll
total 0
lrwxrwxrwx 1 root root 27 Sep  7 05:34 cfgen -> /root/test/hare/cfgen/cfgen
lrwxrwxrwx 1 root root 38 Sep  7 05:34 configure -> /root/test/hare/.py3venv/bin/configure
lrwxrwxrwx 1 root root 46 Sep  7 05:34 dhall -> /root/test/hare/vendor/dhall-bin/current/dhall
lrwxrwxrwx 1 root root 54 Sep  7 05:34 dhall-to-json -> /root/test/hare/vendor/dhall-bin/current/dhall-to-json
lrwxrwxrwx 1 root root 54 Sep  7 05:34 dhall-to-yaml -> /root/test/hare/vendor/dhall-bin/current/dhall-to-yaml
lrwxrwxrwx 1 root root 45 Sep  7 05:34 generate_sysconf -> /root/test/hare/.py3venv/bin/generate_sysconf
lrwxrwxrwx 1 root root 39 Sep  7 05:34 hare_setup -> /root/test/hare/.py3venv/bin/hare_setup
lrwxrwxrwx 1 root root 32 Sep  7 05:34 hax -> /root/test/hare/.py3venv/bin/hax
lrwxrwxrwx 1 root root 20 Sep  7 05:34 hctl -> /root/test/hare/hctl
lrwxrwxrwx 1 root root 54 Sep  7 05:34 json-to-dhall -> /root/test/hare/vendor/dhall-bin/current/json-to-dhall
lrwxrwxrwx 1 root root 30 Sep  7 05:34 q -> /root/test/hare/.py3venv/bin/q
lrwxrwxrwx 1 root root 54 Sep  7 05:34 yaml-to-dhall -> /root/test/hare/vendor/dhall-bin/current/yaml-to-dhall
```

Here is the output of the config stage of the provisioner:
```
[root@ssc-vm-g2-rhev4-0399 ~]# /opt/seagate/cortx/hare/bin/hare_setup config --config 'json:///root/hare.config.conf.tmpl.1-node' --file '/var/lib/hare/cluster.yaml'
```
Files generated successfully:
```
[root@ssc-vm-g2-rhev4-0399 bin]# cd /etc/sysconfig

[root@ssc-vm-g2-rhev4-0399 sysconfig]# ls
m0d-0x7200000000000001:0x19  m0d-0x7200000000000001:0x28  m0d-0x7200000000000001:0xa  s3server-0x7200000000000001:0x2b
```

knekrasov commented on hax/helper/generate_sysconf.py:
```diff
@@ -0,0 +1,159 @@
+#!/usr/bin/env python3
```
Shebang is not needed - this is not an executable script per se.

supriyachavan4398 replied:
> Shebang is not needed - this is not an executable script per se.

Addressed.

knekrasov commented on hax/helper/generate_sysconf.py:
```diff
@@ -0,0 +1,159 @@
+                'matches the one stored in the Consul KV.')
+        generator.generate_sysconfig(IDs)
+    except Exception as e:
+        logging.error('Exiting due to a failure: %s', e)
```
I think it makes sense to exit with non-zero code if an exception is catched here.

supriyachavan4398 replied:
> I think it makes sense to exit with non-zero code if an exception is catched here.

Addressed.

knekrasov reviewed (commented):
Please add at least some basic docstrings and please also make sure that the resulting `generate_sysconf` script works after `make install` (i.e. make sure hax imports are not broken)

mssawant commented:
@Shreya-18, this looks good but Hare does the similar thing in utils/update-consul-conf script, update-consul-conf script fetches the respective IDs from Consul, creates /etc/sysconfig/{m0d-*, s3server-*} files and also updates consul-server-conf.json.

Now, we are duplicating the functionality. We need to make sure to re-use the code. Having a python script is fine but either it should be possible to invoke this script from `update-consul-conf` or `update-consul-conf functionality must be completely moved to this script.

You can do the following,
- Change this script into a module returning respective fids. Implement a wrapper script with command options to return fids for a given Motr process (I think similar functionality is provided by `hctl fetch-fids` ?)
- We can then invoke this script command that returns the fids from update-consul-conf script which will use those fids to update consul-server-conf.json.
- Or move complete `update-consul-conf` functionality to your new module which fetches fids, creates sysconfig files and updates `consul-server-conf.json`.

What do you think?

knekrasov commented:
@mssawant 
>this looks good but Hare does the similar thing in utils/update-consul-conf script, update-consul-conf script fetches the respective IDs from Consul, creates /etc/sysconfig/{m0d-, s3server-} files and also updates consul-server-conf.json.

> Now, we are duplicating the functionality. We need to make sure to re-use the code.

In fact, this was my idea to have this kind of duplication. The reasoning was as follows:
1. In ideal situation we need to have all the code reused and any duplication should be avoided.
2. But this means that quite many pieces of hare-bootstrap and cfgen functionality should be refactored (in addition, hare-bootstrap and update-consul-conf scripts should be translated into Python). This is risky (regressions are possible given this functionality is not covered with auto tests - so we even may not notice that something has been broken) and enlarges the scope of work quite significantly (which is not so good also in the light of the tight schedule).

So my advice was to not to be afraid to duplicate the code AND plan the proper refactoring in the future (when the deadlines become not so agressive).

BTW when implementing #1775 I had the same idea in mind. I didn't try to refactor hare-bootstrap but concentrated on MVP instead.

Shreya-18 commented:
@mssawant 
> @Shreya-18, this looks good but Hare does the similar thing in utils/update-consul-conf script
> Now, we are duplicating the functionality. We need to make sure to re-use the code.
> 
**Approach 1**
> * Change this script into a module returning respective fids. Implement a wrapper script with command options to return fids for a given Motr process (I think similar functionality is provided by `hctl fetch-fids` ?)
> * We can then invoke this script command that returns the fids from update-consul-conf script which will use those fids to update consul-server-conf.json.

Here, we are not just fetching fids, but also endpoints, meta_data, profile_fids. So we need to access consul kv anyway, as `hctl fetch-fids` returns only fids.

New Module 
```
Get_fid(), Get_ep(fid), get_md(), get_Profile fid(), ...
Generate_sysconfig() – generates sysconfig files.
```
Wrapper script – 
```
command options 
--fid, --endpoint, --metadata, --prof-fid, etc
```

Update-Consul-conf
```
1.	Use wrapper script to initialize all variables. (used by dry run)
2.	Use the above variables for writing consul-server-conf.json file.
```
Config stage
```
1.	Call module abc.generate_sysconf() 
```
**Approach 2**
> * Or move complete `update-consul-conf` functionality to your new module which fetches fids, creates sysconfig files, and updates `consul-server-conf.json`.
> 

bootstrap-node and few other scripts uses the `dry-run` functionality (source update-consul-conf --dry-run ). This we need to handle. 

- bash needs to call python functions of new module OR
- hctl fetch-fids ex. (`hctl fetch-fids -s ioservice | jq -r '.[].fid'`)

Also, Can’t we simply call `update-consul-conf` script from the config stage? (We can update its functionality to simply generate the only sysconfig files- if needed) – cons: things remain in bash.
Plus, we need to be extra cautious to avoid any regression as @knekrasov mentioned.

supriyachavan4398 commented:
> Please add at least some basic docstrings and please also make sure that the resulting `generate_sysconf` script works after `make install` (i.e. make sure hax imports are not broken)

After running `make install` :
```
--> Altering generated executables to make Python imports work
    Fixing /opt/seagate/cortx/hare/bin/hare_setup
    Fixing /opt/seagate/cortx/hare/bin/configure
    Fixing /opt/seagate/cortx/hare/bin/generate_sysconf
```

and the resulting `generate_sysconf` script works fine.
```
[root@ssc-vm-g2-rhev4-0399 bin]# ./generate_sysconf
2021-09-08 04:54:31,054 [DEBUG] Starting new HTTP connection (1): 127.0.0.1:8500
2021-09-08 04:54:31,057 [DEBUG] http://127.0.0.1:8500 "GET /v1/kv/m0conf/nodes?recurse=1 HTTP/1.1" 200 1546
2021-09-08 04:54:31,064 [DEBUG] http://127.0.0.1:8500 "GET /v1/kv/m0conf/nodes?recurse=1 HTTP/1.1" 200 1546
2021-09-08 04:54:31,070 [DEBUG] http://127.0.0.1:8500 "GET /v1/kv/m0conf/nodes?recurse=1 HTTP/1.1" 200 1546
2021-09-08 04:54:31,075 [DEBUG] http://127.0.0.1:8500 "GET /v1/kv/m0conf/nodes?recurse=1 HTTP/1.1" 200 1546
2021-09-08 04:54:31,079 [DEBUG] http://127.0.0.1:8500 "GET /v1/kv/m0conf/nodes/srvnode-1.data.private/processes/7/endpoint HTTP/1.1" 200 173
2021-09-08 04:54:31,082 [DEBUG] http://127.0.0.1:8500 "GET /v1/kv/m0conf/nodes/srvnode-1.data.private/processes/40/endpoint HTTP/1.1" 200 174
2021-09-08 04:54:31,085 [DEBUG] http://127.0.0.1:8500 "GET /v1/kv/m0conf/nodes/srvnode-1.data.private/processes/10/endpoint HTTP/1.1" 200 174
2021-09-08 04:54:31,088 [DEBUG] http://127.0.0.1:8500 "GET /v1/kv/m0conf/nodes/srvnode-1.data.private/processes/10/meta_data HTTP/1.1" 200 183
2021-09-08 04:54:31,092 [DEBUG] http://127.0.0.1:8500 "GET /v1/kv/m0conf/nodes/srvnode-1.data.private/processes/25/endpoint HTTP/1.1" 200 174
2021-09-08 04:54:31,094 [DEBUG] http://127.0.0.1:8500 "GET /v1/kv/m0conf/nodes/srvnode-1.data.private/processes/25/meta_data HTTP/1.1" 200 183
2021-09-08 04:54:31,097 [DEBUG] http://127.0.0.1:8500 "GET /v1/kv/m0conf/profiles/?keys=True HTTP/1.1" 200 43
2021-09-08 04:54:31,100 [DEBUG] http://127.0.0.1:8500 "GET /v1/kv/m0conf/nodes/srvnode-1.data.private/processes/43/endpoint HTTP/1.1" 200 174
```

mssawant commented:
Hi @Shreya-18, can you post the generated consul-server-conf.json file?
Will try to test the changes with containers.

Shreya-18 commented:
> Hi @Shreya-18, can you post the generated consul-server-conf.json file?
> Will try to test the changes with containers.

Here's it
```
{
  "server": true,
  "leave_on_terminate": true,
  "watches": [
    {
      "type": "key",
      "key": "leader",
      "args": [
        "/opt/seagate/cortx/hare/libexec/elect-rc-leader"
      ]
    },
    {
      "type": "keyprefix",
      "prefix": "bq/",
      "handler_type": "http",
      "http_handler_config": {
        "path": "http://10.230.246.59:8008/watcher/bq",
        "method": "POST",
        "timeout": "10s"
      }
    },
    {
      "type": "service",
      "service": "confd",
      "handler_type": "http",
      "http_handler_config": {
        "path": "http://10.230.246.59:8008",
        "method": "POST",
        "timeout": "10s"
      }
    },
    {
      "type": "service",
      "service": "confd",
      "args": [
        "/opt/seagate/cortx/hare/libexec/consul-watch-handler"
      ]
    },
    {
      "type": "service",
      "service": "ios",
      "handler_type": "http",
      "http_handler_config": {
        "path": "http://10.230.246.59:8008",
        "method": "POST",
        "timeout": "10s"
      }
    },
    {
      "type": "service",
      "service": "ios",
      "args": [
        "/opt/seagate/cortx/hare/libexec/consul-watch-handler"
      ]
    },
    {
      "type": "service",
      "service": "s3service",
      "handler_type": "http",
      "http_handler_config": {
        "path": "http://10.230.246.59:8008",
        "method": "POST",
        "timeout": "10s"
      }
    },
    {
      "type": "service",
      "service": "s3service",
      "args": [
        "/opt/seagate/cortx/hare/libexec/consul-watch-handler"
      ]
    }
  ],
  "enable_local_script_checks": true,
  "services": [
    {
      "id": "6",
      "name": "hax",
      "address": "10.230.246.59@tcp:12345:1",
      "port": 1,
      "checks": [
        {
          "args": [
            "/opt/seagate/cortx/hare/libexec/check-service",
            "--hax"
          ],
          "interval": "1s",
          "status": "warning"
        }
      ]
    },
    {
      "id": "9",
      "name": "confd",
      "address": "10.230.246.59@tcp:12345:2",
      "port": 1,
      "checks": [
        {
          "args": [
            "/opt/seagate/cortx/hare/libexec/check-service",
            "--fid",
            "0x7200000000000001:0x9"
          ],
          "interval": "1s",
          "status": "warning"
        }
      ]
    },
    {
      "id": "12",
      "name": "ios",
      "address": "10.230.246.59@tcp:12345:2",
      "port": 2,
      "checks": [
        {
          "args": [
            "/opt/seagate/cortx/hare/libexec/check-service",
            "--fid",
            "0x7200000000000001:0xc"
          ],
          "interval": "1s",
          "status": "warning"
        }
      ]
    }
  ]
}
```

Shreya-18 commented:
Wrapper script
```
$ /opt/seagate/cortx/hare/bin/update_conf -h
usage: update_conf [OPTION]

Generates configuration files for consul-agent, motr and s3 services.

optional arguments:
  -h, --help     show this help message and exit
  --sysconf      Only generate system configuration files for motr and s3
                 services.
  --consul_conf  Only update consul agent configuration file.
  --all          Generates sysconfig files for motr, s3 services and also
                 updates consul-agent conf files.
  --fid FID      Returns fids of the service for the current nodes in cluster.
                 List of service fids - "hax", "confd", "ios", "s3".

root@ssc-vm-g2-rhev4-0399.colo.seagate.com:/root/test/hare (t2)
$ /opt/seagate/cortx/hare/bin/update_conf --fid hax
[0x7200000000000001:0x7]

root@ssc-vm-g2-rhev4-0399.colo.seagate.com:/root/test/hare (t2)
$ /opt/seagate/cortx/hare/bin/update_conf --fid ios
[0x7200000000000001:0xd, 0x7200000000000001:0x1e]
```

mssawant commented on hax/helper/generate_sysconf.py:
```diff
@@ -0,0 +1,255 @@
+# Copyright (c) 2020 Seagate Technology LLC and/or its Affiliates
```
2021 please.

Shreya-18 replied:
done

mssawant commented on hax/helper/generate_sysconf.py:
```diff
@@ -0,0 +1,255 @@
+
+
+def get_node_name():
+    with open('/var/lib/hare/node-name') as f:
```
/var/lib/hare/node-name is created by `configure` script. We need to be sure that when generate_sysconf.py is invoked this file is present. I think we better take node-name as a input to the respective functions.

Shreya-18 replied:
took as a parameter. done

mssawant commented on hax/helper/generate_sysconf.py:
```diff
@@ -0,0 +1,255 @@
+
+    def __init__(self):
+        self.kv = KVAdapter()
+        self.sysconfig_dir = '/etc/sysconfig/'
```
We need to avoid hard coding the directory paths. Let's take it as an input to the Generator constructor.

Shreya-18 replied:
done

mssawant commented on hax/helper/generate_sysconf.py:
```diff
@@ -0,0 +1,255 @@
+    def __init__(self):
+        self.kv = KVAdapter()
+        self.sysconfig_dir = '/etc/sysconfig/'
+        if not os.path.isdir(self.sysconfig_dir):
```
No need to have a default path, make the config dir path mandatory.

Shreya-18 replied:
removed

mssawant commented on hax/helper/generate_sysconf.py:
```diff
@@ -0,0 +1,255 @@
+            self.sysconfig_dir = '/etc/default/'
+
+    @repeat_if_fails()
+    def get_service_ids(self, svc_name: str) -> List[str]:
```
As discussed let's have offline version of these functions too.

Shreya-18 replied:
done, added 2 seperate classes to hold the online and offline versions. Based on kv file input, functions are called.

mssawant commented on hax/helper/update_conf.py:
```diff
@@ -0,0 +1,90 @@
+# Copyright (c) 2020 Seagate Technology LLC and/or its Affiliates
```
2021 please.

Shreya-18 replied:
done

mssawant commented on hax/helper/generate_sysconf.py:
```diff
@@ -0,0 +1,255 @@
+        return Service(id=id, name=name, address=addr, port=port,
+                       checks=[checks])
+
+    def update_consul_conf(self, conf_dir='/var/lib/hare'):
```
Make conf_dir a mandatory parameter and don't set a default value. Caller must decide on passing the value.

Shreya-18 replied:
yes done

mssawant commented on hax/helper/generate_sysconf.py:
```diff
@@ -0,0 +1,255 @@
+
+        Examples of the key that will match:
+            key = m0conf/nodes/srvnode-1/processes/10/endpoint
+            value = 10.230.246.59@tcp:12345:2:1
```
The endpoint format will change with lifabric.

Shreya-18 replied:
yes done

mssawant commented on hax/helper/generate_sysconf.py:
```diff
@@ -0,0 +1,255 @@
+    def get_service_port(self, ep: str) -> int:
+        return int(ep.rsplit('@', 1)[1])
+
+    def prepare_svc(self, id: str, name: str):
```
Just to make codacy happy, please rename the id to may be svc_id or fid?

Shreya-18 replied:
done

mssawant commented on hax/helper/generate_sysconf.py:
```diff
@@ -0,0 +1,255 @@
+        contents = (f'MOTR_M0D_EP={ep}\n'
+                    f'MOTR_HA_EP={hax_ep}\n'
+                    f'MOTR_PROCESS_FID={fid}\n'
+                    'MOTR_CONF_XC=/etc/motr/confd.xc\n')
```
Let's take motr conf dir path as an input to this function or to the Generator class itself. Avoid hard coding the paths.

Shreya-18 replied:
done

knekrasov commented on hax/helper/generate_sysconf.py:
```diff
@@ -0,0 +1,347 @@
+    def __init__(self, node: str, kv_file: Optional[str] = None):
+        # Based on the kv_file input, we choose whether to fetch data from
+        # kv file or consul kv.
+        self.provider = KVFile(kv_file=kv_file,
+                               node=node) if kv_file else ConsulKV(node)
```
[Optional]
I would add a type hint for `self.provider` (either by making KVFile and ConsulKV to have the same base class or by using [`Union[]`](https://docs.python.org/3/library/typing.html#typing.Union))

Shreya-18 replied:
done.. used Union

knekrasov commented on hax/helper/generate_sysconf.py:
```diff
@@ -0,0 +1,347 @@
+        with fileinput.FileInput(conf_file, inplace=True,
+                                 backup='.bak') as file:
+            for line in file:
+                print(line.replace('http://localhost',
```
It is just printing to the console, isn't it?

Why the function name is `append_ipaddr_to_file` then? I would expect that the function modifies the file somehow but it doesn't happen.

Shreya-18 replied:
Actually, FileInput supports in-place replacing- here it **redirects standard output to the input file** (first the input file is moved to the backup file), and it is deleted when the file is closed.
ref - https://docs.python.org/3/library/fileinput.html#fileinput.FileInput

knekrasov replied:
It does, but I guess that `file` should be passed as a [parameter to print](https://stackoverflow.com/questions/36571560/directing-print-output-to-a-txt-file) then. Am I wrong?

Shreya-18 replied:
[in-place editing file](https://www.geeksforgeeks.org/python-inplace-editing-using-fileinput/ ) here, stdout is redirected to file within that context manager.

Shreya-18 replied:
And I tested the code.. it seems to work as expected.

knekrasov replied:
I suspected something like that but I didn't manage to find a proof in the official documents. If you have the link that describes the behaviour of context manager + print, please add to the code as a comment. Otherwise it looks too magic.

Shreya-18 replied:
> Optional in-place filtering: if the keyword argument inplace=True is passed to the FileInput constructor, the file is moved to a backup file and standard output is directed to the input file. This makes it possible to write a filter that rewrites its input file in place. 

Basically, the file taken in constructor of FileInput .. the stdout is directed to that file.

Shreya-18 replied:
Done added the link and explanation as comments.

Shreya-18 commented on hax/helper/generate_sysconf.py:
```diff
@@ -0,0 +1,347 @@
+                           f'/services/{svc_name}$')
+        keys = []
+        for key in self.kv_data:
+            match_result = re.match(regex, key['key'])
```
Yes, in Consul its `Key` and `Value` whereas in `consul-kv.json` its `key` and `value`. 
here's the snippet from the file.
```
 {
    "key": "m0conf/profiles/0x7000000000000001:0x60",
    "value": "{\"name\": \"default\", \"pools\": [\"the pool\", null, null]}"
  },

Shreya-18 commented on hax/helper/generate_sysconf.py:
```diff
@@ -0,0 +1,347 @@
+        self.kv_data = self._read_file()
+        self.node = node
+
+    def _read_file(self) -> List:
```
Yes, I'll add the type param. Thanks. And its `List[Dict[str, Any]]`. Seems types are very useful

Shreya-18 replied:
done

Shreya-18 commented on hax/helper/generate_sysconf.py:
```diff
@@ -0,0 +1,347 @@
+    name: str
+    address: str
+    port: int
+    checks: List[Dict]
```
done

mssawant commented:
@Shreya-18, need to print the usage of `update_conf` if executed without any arguments,
```
[root@cortx-hare-motr-prov-pod ~]# /opt/seagate/cortx/hare/bin/update_conf --help
usage: update_conf [OPTION]

Generates configuration files for consul-agent, motr and s3 services.

positional arguments:
  node                  Node-name of the services whose config files are
                        generated.

optional arguments:
  -h, --help            show this help message and exit
  --sysconf SYSCONF, -s SYSCONF
                        Path to generate system configuration files for motr
                        and s3 services.
  --consul-conf CONSUL_CONF, -c CONSUL_CONF
                        Path to hare config directory for updating consul
                        agent configuration file.
  --motr-conf MOTR_CONF, -m MOTR_CONF
                        Path to motr config directory.
  --kv-file KV_FILE     Hare-Motr configuration key values file path.
  --fid FID             Returns fids of the service for the current nodes in
                        cluster. List of service fids - "hax", "confd", "ios",
                        "s3".
[root@cortx-hare-motr-prov-pod ~]# /opt/seagate/cortx/hare/bin/update_conf
usage: update_conf [OPTION]
update_conf: error: the following arguments are required: node
```
Kindly update the name to `update-conf` instead of `update_conf`.

mssawant commented:
Rebased with latest `kubernetes` branch, `hare_setup` `prepare` and `config` stages completed successfully. Was able to start hax as well,
```
[root@cortx-hare-motr-prov-pod ~]# curl localhost:8500/v1/agent/health/service/name/hax
[{"AggregatedStatus":"passing","Service":{"ID":"6","Service":"hax","Tags":[],"Meta":{},"Port":2001,"Address":"inet:tcp:cortx-hare-motr-prov-pod","Weights":{"Passing":1,"Warning":1},"EnableTagOverride":false,"Datacenter":"dc1"},"Checks":[{"Node":"cortx-hare-motr-prov-pod","CheckID":"service:6","Name":"Service 'hax' check","Status":"passing","Notes":"","Output":"  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n                                 Dload  Upload   Total   Spent    Left  Speed\n\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\ncurl: (56) Recv failure: Connection reset by peer\n","ServiceID":"6","ServiceName":"hax","ServiceTags":null,"Type":"","Definition":{"Interval":"0s","Timeout":"0s","DeregisterCriticalServiceAfter":"0s","HTTP":"","Header":null,"Method":"","Body":"","TLSSkipVerify":false,"TCP":""},"CreateIndex":0,"ModifyIndex":0}]}]
```

Shreya-18 commented:
> Kindly update the name to `update-conf` instead of `update_conf`.

Done!!
```
$ /opt/seagate/cortx/hare/bin/update-conf --help
usage: update-conf [OPTION]

Generates configuration files for consul-agent, motr and s3 services.

positional arguments:
  node                  Node-name of the services whose config files are
                        generated.

optional arguments:
  -h, --help            show this help message and exit
  --sysconf SYSCONF, -s SYSCONF
...

Shreya-18 commented:
```$ /opt/seagate/cortx/hare/bin/update-conf
usage: update-conf node [OPTION]
update-conf: error: the following arguments are required: node
```
updated the usage and rebased with the latest Kubernetes branch. :)

mssawant commented:
@Shreya-18,
```
[root@cortx-hare-motr-prov-pod ~]# /opt/seagate/cortx/hare/bin/update_conf --help
usage: update_conf [OPTION]

Generates configuration files for consul-agent, motr and s3 services.

positional arguments:
  node                  Node-name of the services whose config files are
                        generated.

optional arguments:
  -h, --help            show this help message and exit
  --sysconf SYSCONF, -s SYSCONF
                        Path to generate system configuration files for motr
                        and s3 services.
  --consul-conf CONSUL_CONF, -c CONSUL_CONF
                        Path to hare config directory for updating consul
                        agent configuration file.
  --motr-conf MOTR_CONF, -m MOTR_CONF
                        Path to motr config directory.
  --kv-file KV_FILE     Hare-Motr configuration key values file path.
  --fid FID             Returns fids of the service for the current nodes in
                        cluster. List of service fids - "hax", "confd", "ios",
                        "s3".
```
Let's avoid mentioning `consul` explicitly, I think let's rename `--consul-conf` to `--hare-conf` as mentioned in the help "path to hare config dir".
Also, we can avoid `sysconf` option as well, overall I think we just need a config dir path, e.g. `/var/lib/` or `/etc/cortx` or `/etc/cortx/{node-name}` internally we can check if `{config-dir}/hare` exists or `{conf-dir}/hare/consul-server-conf` exists.
Expectations are to have configuration files already present in `{conf-dir}/hare` and `{conf-dir}/hare/consul-server-conf`.
Rename `--motr-conf` to `motr-conf-dir` , `--sysconf` and `--consul-conf` can be replaced with `--confg-dir`.
I think we will also have to add `--s3-conf-dir` option where the s3 sysconfig files will be generated. So ideally all the respective, hare, motr and s3 config dirs will have a corresponding `sysconfig` dir container corresponding files.
Makes sense?

mssawant reviewed (approved):
Looks great! Few more cosmetic changes suggested. Kindly check the same.
No further reviews required from my side. Just check if all the files are created in the respective directory paths.
Thank you!

Shreya-18 commented:
> Let's avoid mentioning `consul` explicitly, I think let's rename `--consul-conf` to `--hare-conf` as mentioned in the help "path to hare config dir".
> Also, we can avoid `sysconf` option as well, overall I think we just need a config dir path, e.g. `/var/lib/` or `/etc/cortx` or `/etc/cortx/{node-name}` internally we can check if `{config-dir}/hare` exists or `{conf-dir}/hare/consul-server-conf` exists.
> Expectations are to have configuration files already present in `{conf-dir}/hare` and `{conf-dir}/hare/consul-server-conf`.
> Rename `--motr-conf` to `motr-conf-dir` , `--sysconf` and `--consul-conf` can be replaced with `--hare-conf-dir`.
> I think we will also have to add `--s3-conf-dir` option where the s3 sysconfig files will be generated. So ideally all the respective, hare, motr and s3 config dirs will have a corresponding `sysconfig` dir containing corresponding files.
> Makes sense?

Done!!

```
$ /opt/seagate/cortx/hare/bin/update-conf -h
usage: update-conf node [OPTION]

Generates configuration files for consul-agent, motr and s3 services.

positional arguments:
  node                  Node-name of the services whose config files are
                        generated.

optional arguments:
  -h, --help            show this help message and exit
  --hare-conf-dir HARE_CONF_DIR
                        Path to hare config directory.
  --motr-conf-dir MOTR_CONF_DIR, -m MOTR_CONF_DIR
                        Path to motr config directory.
  --s3-conf-dir S3_CONF_DIR, -s S3_CONF_DIR
                        Path to S3 config directory.
  --kv-file KV_FILE     Hare-Motr configuration key values file path.
  --fid FID             Returns only the fids of the service for the current
                        nodes in cluster. List of service fids - "hax",
                        "confd", "ios", "s3". No configuration files are
                        created.

$ /opt/seagate/cortx/hare/bin/update-conf ssc-vm-g2-rhev4-0399.colo.seagate.com --hare-conf-dir /var/lib/hare --kv-file /var/lib/hare/consul-kv.json --motr-conf-dir /etc/cortx/motr --s3-conf-dir /etc/cortx/s3
2021-09-21 01:41:31,871 [DEBUG] Writing to file /etc/cortx/motr/sysconfig/m0d-0x7200000000000001:0xa
2021-09-21 01:41:31,873 [DEBUG] Writing to file /etc/cortx/motr/sysconfig/m0d-0x7200000000000001:0xd
2021-09-21 01:41:31,874 [DEBUG] Writing to file /etc/cortx/motr/sysconfig/m0d-0x7200000000000001:0x1e
2021-09-21 01:41:31,876 [DEBUG] Writing to file /var/lib/hare//consul-server-conf/consul-server-conf.json
```
Files are being generated in respective directories.

For fetching fids
```
$ /opt/seagate/cortx/hare/bin/update-conf ssc-vm-g2-rhev4-0399.colo.seagate.com --hare-conf-dir /var/lib/hare --fid ios --kv-file /var/lib/hare/consul-kv.json
[0x7200000000000001:0xd, 0x7200000000000001:0x1e]
